x-pack/plugins/licensing/public/plugin.ts

```diff
@@ -0,0 +1,127 @@
+
+export class Plugin implements CorePlugin<LicensingPluginSetup>, ILicensingPlugin {
+  private http!: CoreSetup['http'];
+  private removeInterceptor!: () => void;
+  private signature = '';
+  public pollingFrequency = DEFAULT_POLLING_FREQUENCY;
+
+  constructor(context: PluginInitializerContext) {}
+
+  private getSession() {
+    const raw = sessionStorage.getItem(LICENSING_SESSION);
+    const signature = sessionStorage.getItem(LICENSING_SESSION_SIGNATURE);
+    const json = raw ? JSON.parse(raw) : {};
+
+    if (signature) {
+      this.signature = signature;
+    }
+
+    return json;
+  }
+
+  private setSession = (license: License) => {
+    sessionStorage.setItem(LICENSING_SESSION, JSON.stringify(license.toObject()));
+
+    if (license.signature) {
+      sessionStorage.setItem(LICENSING_SESSION_SIGNATURE, license.signature);
+    }
+  };
+
+  private clearSession() {
+    sessionStorage.removeItem(LICENSING_SESSION_SIGNATURE);
+    sessionStorage.removeItem(LICENSING_SESSION);
+  }
+
+  private next = async () => {
+    try {
+      const response = await this.http.get(API_ROUTE);
+      const rawLicense = response && response.license;
+      const features = (response && response.features) || {};
+
+      return new License({ plugin: this, license: rawLicense, features });
+    } catch (err) {
+      // Prevent reusing stale license if the fetch operation fails
+      this.clearSession();
+      return new License({ plugin: this, features: {}, error: err });
+    }
+  };
+
+  private intercept(refresher$: BehaviorSubject<boolean>) {
+    this.removeInterceptor = this.http.intercept({
+      response: httpResponse => {
+        const signature = httpResponse.response!.headers.get(SIGNATURE_HEADER) || '';
+
+        if (signature !== this.signature) {
+          this.signature = signature;
+
+          if (!httpResponse.request!.url.includes(API_ROUTE)) {
+            refresher$.next(true);
+          }
+        }
+      },
+    });
+  }
+
+  public sign() {
+    return this.signature;
+  }
+
+  public setup(core: CoreSetup) {
+    this.http = core.http;
+
+    const { license, features = {} } = this.getSession();
+    const refresher$ = new BehaviorSubject<boolean>(true);
+
+    this.intercept(refresher$);
+
+    const initial$ = of(new License({ plugin: this, license, features }));
+    const licenseFetches$ = merge(timer(0, this.pollingFrequency), refresher$).pipe(
+      switchMap(this.next)
+    );
+    const updates$ = merge(initial$, licenseFetches$).pipe(
+      pairwise(),
+      filter(([previous, next]) => hasLicenseInfoChanged(previous, next)),
+      map(([, next]) => next)
+    );
+    const license$ = merge(initial$, updates$).pipe(tap(this.setSession));
+
+    return {
+      license$,
+      refresh() {
+        refresher$.next(true);
+      },
+    };
+  }
+
+  public async start(core: CoreStart) {}
+
+  public stop() {
+    if (this.removeInterceptor) {
```

Comment:
shouldn't we stop all observables?

Reply:
Yeah we should probably follow the `takeUntil(stop$)` pattern used elsewhere. @eliperelman check out the ChromeService for an example.